Add suggested tags to post publish share modal

eoigal opened the pull request:
This PR intends to add a list of suggested tags based on the post content to the post-publish share modal.

The suggested tags will only show for;

* Users with suggested tags in post meta

The suggested tags are added when;

* The post has post content (more than 50 words)
* The site is in an English locale
* The post auto saves **AND** there are no suggested tags saved in post meta
* The user manually saves the post

The project can be found here - p2-pe7F0s-1jr

### Tracks Events

* `calypso_reader_post_publish_add_tags` - Tracks when a user clicks Add tags button which sends API request to save tags to post meta
* `calypso_reader_post_publish_update_suggested_tags` - Tracks when a user changes the tags in the textfield
* `calypso_reader_post_publish_show_suggested_tags` - Tracks when a user is shown the suggested tags

### Design

![tags-1](https://github.com/Automattic/wp-calypso/assets/5560595/47cf8e28-bda6-4922-8b29-721676387df7)

### Result

https://github.com/Automattic/wp-calypso/assets/5560595/b79ff86a-3a07-4b1d-bb6c-1ea1626091c8


### Testing

* Apply the patch code-D128037 on sandbox
* Apply this change to editing-toolkit to your sandbox with `cd apps/editing-toolkit && yarn dev --sync`
* Create a post on a test sandboxed site with more than 50 words. The AI assistant block is useful here.
* Publish the post and you should see the suggested tags in the post publish modal
* To retest, you can switch the post to a draft and publish again to see the post publish modal.

eoigal committed:
Add suggested tags to post publish share modal

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/index.tsx:
```diff
@@ -65,9 +68,10 @@ const SharingModalInner: React.FC = () => {
 		useShouldShowVideoCelebrationModal( isCurrentPostPublished );
 
-	const [ isOpen, setIsOpen ] = useState( false );
+	const [ isOpen, setIsOpen ] = useState( true );
```
I imagine this might be for convenience during development?

In testing this the modal was present as soon as I opened the editor.

eoigal replied:
Yep, this is only here for testing so we can open the post published dialog easily

Addison-Stavlo reviewed (commented):
I asked chatGPT to write up 3 paragraphs about "Spam Musubi" and used this as my post:

> Spam musubi is a popular and beloved Hawaiian snack that beautifully combines the flavors of Hawaii and its strong ties to Asian cuisine. This delectable treat consists of a slice of grilled or pan-fried Spam (a canned meat product) placed on a block of seasoned sushi rice and wrapped in a strip of nori (seaweed). The result is a delightful fusion of textures and tastes, with the crisp nori, savory Spam, and slightly sweet, vinegared rice coming together in perfect harmony. Although it may not be the most traditional or sophisticated dish, spam musubi has earned a special place in the hearts of Hawaiians and visitors alike, offering a convenient, portable, and flavorful option for those on the go.

> The origins of spam musubi can be traced back to the mid-20th century when Spam became a staple in Hawaii due to its long shelf life and affordability. Local residents quickly embraced it and integrated it into their culinary repertoire. The idea of wrapping Spam with rice and nori is reminiscent of sushi, but it represents a unique Hawaiian adaptation of this Japanese tradition. Today, spam musubi is widely available at convenience stores, local eateries, and food stalls across Hawaii, and it has also gained popularity in other parts of the world, thanks to its delicious simplicity.

> Spam musubi is not only a quick and satisfying snack, but it also serves as a symbol of the cultural diversity and culinary creativity that define Hawaiian cuisine. Its fusion of flavors and influences is a testament to the island's rich history, where different culinary traditions have come together to create something truly unique and delicious. Whether enjoyed as a snack during a day at the beach or as part of a traditional Hawaiian plate lunch, spam musubi continues to be a beloved and enduring part of the Hawaiian culinary landscape.

The result had no suggestions:

<img width="1104" alt="Screenshot 2023-11-09 at 9 54 01 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/9be5fcbf-73f0-4eb8-841a-d0a3f2d16e9b">

It seems like the changes were present, as I do see the copy change to "TEST" as well as the modal opening as soon as we enter the editor.

eoigal commented:
Sorry missed committing an update to code-D128037.

Addison-Stavlo commented:
Nice, im seeing the same thing! 
<img width="1421" alt="Screenshot 2023-11-09 at 1 31 39 PM" src="https://github.com/Automattic/wp-calypso/assets/28742426/2f70fec1-149f-419c-86a0-e7e12a7b2da9">

The modal has become a bit wide and the "add these tags" button isnt working on my end yet, but i realize this is still in progress.

The results seem decent! I really like how we can add more tags via this input as well. Even if the relevance of the recommendations is spotty, it really increases the awareness about them and gives users a new way to input them that isnt as hidden as the general sidebar area they may have not found yet. Im liking where this is going.

eoigal commented:
Added code-D129459 to add an endpoint to allow this PR to save tags to the post

eoigal committed:
add method to post tags to be save with post

eoigal committed:
Update modal add tags button to save tags to post

eoigal committed:
remove unused style

eoigal committed:
remove unused state vars

eoigal committed:
Update to create snackbar notice when saved tags and to record events in tracks

eoigal committed:
Tidy up transition between add tags to img

eoigal committed:
Fix some Typescript errors found in unit tests

eoigal committed:
fix more ts errors

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/index.tsx:
```diff
@@ -65,9 +71,12 @@ const SharingModalInner: React.FC = () => {
 	const closeModal = () => setIsOpen( false );
 	const { createNotice } = useDispatch( noticesStore );
+	const [ shouldShowSuggestedTags, setShouldShowSuggestedTags ] = React.useState(
+		postMeta?.reader_suggested_tags?.length > 0
```
Still to test, but reading through this stands out to me as potentially buggy.

If we set the initial state based on this, then say for the case of a new post `shouldShowSuggestedTags` would be set as `false` when first loaded and we have no other effects that toggle this to `true` after saves may update this post meta value to contain suggestions? So the SuggestedTags component would never render since this value would still be false.

Alternatively, we could assume `true` from the beginning. Once the modal toggles to be opened and the `SuggestedTags` component mounts, we could use its check for the post meta there to toggle this to `false` if there are no suggestions present. That would be evaluating the presence of suggested tags at the time of publish as opposed to at the time of opening the editor.

eoigal replied:
Thanks for highlighting this, I can see what you mean that it could (and will) lead to a buggy experience. I've updated to your suggestion and it works well in my testing.

Addison-Stavlo commented:
🤔 Im having suggested tags show up for the posts i tested a couple weeks ago. But for new posts that I am creating today, I cannot get the suggested tags to show up (even using the same content), after manually saving, reloading, etc.

Addison-Stavlo commented:
Ok, after you fixed the non-english locale bug I can confirm this works given the testing instructions around reloading:

> Create a post on a test sandboxed site, make sure to manually save the post - this should add suggested tags to the post meta
Reload the editor - the share modal should show a list of suggested tags like the image above

Although Im uncertain about a non-reload / single session publishing flow as it doesn't seem to work. Go into the editor, add some content, manually save, wait a bit more, publish post -> no suggested tags in the same cases they would be present after reloading after saving.

1. Its probably not working due to [what I pointed out above](https://github.com/Automattic/wp-calypso/pull/84020#discussion_r1408075042).
2. Im curious if we fixed that if it would still work? Will the editor get the new post_meta that was added on the backend without a reload in the first place?

roo2 commented:
I'm seeing the sharing modal pop up when I first edit a new post 

<img width="1002" alt="Screen Shot 2023-11-29 at 4 07 31 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/c6af77f9-b694-41b1-ba24-7e573bbb6e44">

roo2 commented:
hmmm what's the plan to get this working so that we don't have to reload the editor and can see recommendations first time we publish? 

This is really a huge improvement, nice work so far!

roo2 commented:
one minor UI thing, I think there is a smaller standard text style to use for that grey text 
<img width="396" alt="Screen Shot 2023-11-29 at 5 24 35 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/08abe50a-01b6-4b8d-9b0a-43f0a23cc1c8">

I think it might be 12px? just trying to find other examples, but I'm sure it shouldn't be the same as the other text.

roo2 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/use-add-tags-to-post.ts:
```diff
@@ -0,0 +1,28 @@
+			onSaveTags( result.added_tags ?? 0 );
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Error: Unable to add tags. Reason: %s', error );
```
This doesn't format the error which is an object, you could JSON.stringify or log error.message

roo2 replied:
also, it should ideally display an error notification to the user

eoigal replied:
I have tried to handle this case a bit better, I now return number of tags added - if error caught, this will be `0` - we then show notice that 'No Tags Added' which I feel is enough here.

eoigal committed:
Return number of tags added so we can handle case where none have been added

eoigal committed:
Add track events and return warning when no tags were added

eoigal committed:
Revert opening post publish by default when loading editor

eoigal commented:
> one minor UI thing, I think there is a smaller standard text style to use for that grey text
> I think it might be 12px? just trying to find other examples, but I'm sure it shouldn't be the same as the other text.

The font size should be the same as what we use in sidebar, which is inline with Dave's designs too from what I can tell;

<img width="297" alt="Screenshot 2023-11-29 at 14 12 00" src="https://github.com/Automattic/wp-calypso/assets/5560595/c6ad7f91-832f-4751-9fe4-cf92b3dc360b">

eoigal committed:
add eventprop

eoigal committed:
update eventprop name

eoigal commented:
Removed test change that opened the post publish modal when loading editor, now you need to publish a post to see it.

I've updated the test instructions to explain how to test it with this change.

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/suggested-tags.tsx:
```diff
@@ -0,0 +1,116 @@
+		[]
+	);
+	const { createNotice } = useDispatch( noticesStore );
+	const origSuggestedTags = postMeta?.reader_suggested_tags
```
Should we `setShouldShowSelectedTags( false )` if `origSuggestedTags` has 0 length?

Right now we start it as true, and set it as false once we click to add the tags. But in the case that there are no tags to suggest it seems like it would remain open with no suggestions?

eoigal replied:
Good idea! It should be false if no suggested tags are found.

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/suggested-tags.tsx:
```diff
@@ -0,0 +1,116 @@
+		/>
+	);
+
+	recordTracksEvent( 'calypso_reader_post_publish_show_suggested_tags', {
+		number_of_original_suggested_tags: origSuggestedTags?.length ?? 0,
+	} );
```
We could wrap this in an on-mount `useEffect` to ensure it only fires once. We could combine this with my other comment just above about closing this component if there are no tags to suggest. The useEffect on mount could be good for both of these items, if there are no tags to suggest close the component and don't send the tracks event, otherwise send this event.

eoigal replied:
Thanks, that's really good suggestion

Addison-Stavlo reviewed (commented):
The single session flow tests well for me now! 🎉 

This is working well! A couple suggestions in comments below but we seem really close here.

<hr />

Aside from all that, I found something a little funky that I think is outside of the scope of this PR, but worth noting as I stumbled upon it.

After adding tags from the modal successfully, I went to look at the tags section in the sidebar (really just to check another interface to see if they had been added, i wouldn't expect users to do this much). The tags section in the sidebar was still empty as if the post had no tags:

<img width="388" alt="Screenshot 2023-11-29 at 11 58 43 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/7685bed7-f677-4473-a44f-951bf3988333">

But reloading the editor it does pick up the tags:
<img width="294" alt="Screenshot 2023-11-29 at 11 59 16 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/88620e9e-09af-48c0-8a07-62a72136b7b3">

I think its a minor concern and a bit of an edge case. I wouldn't expect people to stay in the editor and look at these items after publishing, and its all there and fine if they ever come back to the post later. wdyt?

eoigal committed:
use useEffect to only call tracks event once

eoigal committed:
use useEffect to not show suggested tags if no suggested tags found

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/sharing-modal/suggested-tags.tsx:
```diff
@@ -0,0 +1,120 @@
+
+	useEffect( () => {
+		if ( origSuggestedTags?.length === 0 ) {
+			props.setShouldShowSuggestedTags( false );
```
DO you think its worth having another tracking event here to signify that we opened this and had no tags to suggest? Or is that redundant with the backend logstash items you setup?

eoigal replied:
I was in 2 minds about it. I was thinking we could figure that out but might be simpler to add it.

eoigal replied:
Ok added new track event that will fire if no suggested tags are found and the site is on a English locale. It should be interesting to track to see if we have many sites that are falling into this category - it would suggest that the post content is too small to be included or perhaps something else that we need to investigate.

eoigal committed:
add tracks event to track when no suggested tags found on en site

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9809596

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @eoigal for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9809596).
